core/src/main/java/com/oath/oak/NativeAllocator/OakNativeMemoryAllocator.java

```diff
@@ -142,43 +156,53 @@ public Slice allocateSlice(int size) {
     @Override
     public void free(ByteBuffer bb) {
         allocated.addAndGet(-(bb.remaining()));
-        if (stats != null) stats.release(bb);
-        freeList.add(
-            new AbstractMap.SimpleImmutableEntry<>(
-                freeCounter.getAndIncrement(), new Slice(INVALID_BLOCK_ID,bb)));
+        if (stats != null) {
+            stats.release(bb);
+        }
+        freeList.add(new FreeChuck(freeCounter.getAndIncrement(), bb.remaining(), new Slice(INVALID_BLOCK_ID, bb)));
     }
 
-  public void freeSlice(Slice slice) {
-    allocated.addAndGet(-(slice.getByteBuffer().remaining()));
-    if (stats != null) stats.release(slice.getByteBuffer());
-    freeList.add(new AbstractMap.SimpleImmutableEntry<>(freeCounter.getAndIncrement(), slice));
-  }
+    @Override
+    public void freeSlice(Slice slice) {
+        allocated.addAndGet(-(slice.getByteBuffer().remaining()));
+        if (stats != null) {
+            stats.release(slice.getByteBuffer());
+        }
+        freeList.add(new FreeChuck(freeCounter.getAndIncrement(), slice.getByteBuffer().remaining(), slice));
```

Comment:
I am still worried from where do we get this ByteBuffer and can it's position be not 0? If the ByteBuffer is specially made only for release, please comment about this...

Reply:
The buffer's position is usually not 0

Reply:
From where do we get this BB?

Reply:
From the user, when it frees its buffer.